messaging/src/main/java/org/axonframework/commandhandling/gateway/DefaultReactorCommandGateway.java

```diff
@@ -0,0 +1,252 @@
+ * @since 4.4
+ */
+public class DefaultReactorCommandGateway implements ReactorCommandGateway {
+
+    private final CommandBus commandBus;
+    private final RetryScheduler retryScheduler;
+    private final List<ReactorMessageDispatchInterceptor<CommandMessage<?>>> dispatchInterceptors;
+    private final List<ReactorResultHandlerInterceptor<CommandMessage<?>, CommandResultMessage<?>>> resultInterceptors;
+
+    /**
+     * Creates an instance of {@link DefaultReactorCommandGateway} based on the fields contained in the {@link
+     * Builder}.
+     * <p>
+     * Will assert that the {@link CommandBus} is not {@code null} and throws an {@link AxonConfigurationException} if
+     * it is.
+     * </p>
+     *
+     * @param builder the {@link Builder} used to instantiate a {@link DefaultReactorCommandGateway} instance
+     */
+    protected DefaultReactorCommandGateway(Builder builder) {
+        builder.validate();
+        this.commandBus = builder.commandBus;
+        this.retryScheduler = builder.retryScheduler;
+        this.dispatchInterceptors = builder.dispatchInterceptors;
+        this.resultInterceptors = builder.resultInterceptors;
+    }
+
+    /**
+     * Instantiate a Builder to be able to create a {@link DefaultReactorCommandGateway}.
+     * <p>
+     * The {@code dispatchInterceptors} are defaulted to an empty list.
+     * The {@code resultHandlerInterceptors} are defaulted to an empty list.
+     * The {@link CommandBus} is a <b>hard requirement</b> and as such should be provided.
+     * </p>
+     *
+     * @return a Builder to be able to create a {@link DefaultReactorCommandGateway}
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public <R> Mono<R> send(Object command) {
+        //noinspection unchecked
+        return Mono.<CommandMessage<?>>just(GenericCommandMessage.asCommandMessage(command))
+                .transform(this::processCommandInterceptors)
+                .flatMap(this::dispatchCommand)
+                .flatMap(this::processResultsInterceptors)
+                .transform(this::getPayload);
+    }
+
+    @Override
+    public Registration registerDispatchInterceptor(ReactorMessageDispatchInterceptor<CommandMessage<?>> interceptor) {
+        dispatchInterceptors.add(interceptor);
+        return () -> dispatchInterceptors.remove(interceptor);
+    }
+
+    @Override
+    public Registration registerResultHandlerInterceptor(
+            ReactorResultHandlerInterceptor<CommandMessage<?>, CommandResultMessage<?>> interceptor) {
+        resultInterceptors.add(interceptor);
+        return () -> resultInterceptors.remove(interceptor);
+    }
+
+    private Mono<CommandMessage<?>> processCommandInterceptors(Mono<CommandMessage<?>> commandMessage) {
+        return Flux.fromIterable(dispatchInterceptors)
+                   .reduce(commandMessage, (command, interceptor) -> interceptor.intercept(command))
+                   .flatMap(Mono::from);
```

Comment:
Not sure I follow how you'd use `Function.identity()` in this case. Could you provide an example?

Reply:
Since at this point when `flatMap` is called, the variable is of type `Mono<Mono<CommandMessage<?>>>`. Unless there's a special reason to use `Mono.from` to create another `Mono` from a `Mono` (wrapping inner one), it can be simplified to `flatMap(it -> it)` to use the already existing inner `Mono` instance directly. In this case, `Function.identity()` would replace `it -> it` lambda expression, same as [here](https://github.com/AxonFramework/AxonFramework/pull/1337/commits/1a306018939b35904b579aeaa3b541a3c09e456d)